File: project/6.06_クラスと修飾子/py/py_ex6_06_3/src/ex6_06_3.py
'''
Created on 2023/05/10

@author: Keroichi.T
'''

# 抽象クラスを作成するときはABCライブラリが必要
from abc import ABCMeta, abstractmethod

# 抽象クラス
class Shape(metaclass=ABCMeta):
    @abstractmethod
    def display(self):
        pass
    
    @abstractmethod
    #「->」は関数アノテーションという
    def getName(self) -> str:
        pass

class Circle( Shape ):
    def display(self):
        print( "Circle" )
        
    def getName(self)->str:
        return type(self)

class Rectangle( Shape ):
    def display(self):
        print( "Rectangle" )

    def getName(self)->str:
        # Eclipseは関数アノテーションを無視しても怒られない
        return int(10)


def main():
    # もちろん抽象クラスのインスタンス化はエラー
    # s = Shape()
    s = Circle()
    s.display()
    print(s.getName())
    
    s = Rectangle()
    s.display()
    print(type(s.getName()))

if __name__ == '__main__':
    main()

File: project/6.06_クラスと修飾子/java/java_ex6_06_3/src/java_ex6_06_3/Ex6_06_3.java
package java_ex6_06_3;

abstract class Shape {
	void display() {}
}

class Circle extends Shape {
	void display() {
		System.out.println( "Circle" );
	}
}

class Rectangle extends Shape {
	void display() {
		System.out.println( "Rectangle" );
	}
}

class Triangle extends Shape {
	void display() {
		System.out.println( "Triangle" );
	}
}

class Trapezoid extends Shape {
	void display() {
		System.out.println( "Trapezoid" );
	}
}

class Rhombus extends Shape {
	void display() {
		System.out.println( "Rhombus" );
	}
}

public class Ex6_06_3 {

	public static void main(String[] args) {
		Shape s = new Circle();
		s.display();
		s = new Rectangle();
		s.display();
		s = new Triangle();
		s.display();
		
		s = new Trapezoid();
		s.display();
		s = new Rhombus();
		s.display();
	}

}
